src/openrct2-ui/drawing/engines/opengl/TextureCache.cpp

```diff
@@ -23,38 +23,72 @@
 
 #include <openrct2/drawing/drawing.h>
 
+constexpr uint32 UNUSED_INDEX = 0xFFFFFFFF;
+
+TextureCache::TextureCache()
+{
+    std::fill(_indexMap.begin(), _indexMap.end(), UNUSED_INDEX);
+}
+
 TextureCache::~TextureCache()
 {
     FreeTextures();
 }
 
 void TextureCache::InvalidateImage(uint32 image)
 {
-    auto kvp = _imageTextureMap.find(image);
-    if (kvp != _imageTextureMap.end())
+    uint32 index = _indexMap[image];
+    if (index == UNUSED_INDEX)
+        return;
+
+    CachedTextureInfo& elem = _textureCache.at(index);
+
+    _atlases[elem.index].Free(elem);
+    _indexMap[image] = UNUSED_INDEX;
+
+    if (index == _textureCache.size() - 1)
     {
-        _atlases[kvp->second.index].Free(kvp->second);
-        _imageTextureMap.erase(kvp);
+        // Last element can be popped back.
+        _textureCache.pop_back();
+    }
+    else
+    {
+        // Swap last element with element to erase and then pop back.
+        CachedTextureInfo& last = _textureCache.back();
+
+        // Move last to current.
+        elem = last;
+
+        // Change index for moved element.
+        _indexMap[last.image] = index;
+
+        _textureCache.pop_back();
     }
 }
 
-const CachedTextureInfo* TextureCache::GetOrLoadImageTexture(uint32 image)
+const CachedTextureInfo& TextureCache::GetOrLoadImageTexture(uint32 image)
 {
     image &= 0x7FFFF;
 
-    auto kvp = _imageTextureMap.find(image);
-    if (kvp != _imageTextureMap.end())
+    uint32 index = _indexMap[image];
+    if (index != UNUSED_INDEX)
     {
-        return &kvp->second;
+        return _textureCache[index];
     }
 
-    auto cacheInfo = LoadImageTexture(image);
-    auto cacheItr = _imageTextureMap.insert(std::make_pair(image, cacheInfo));
+    index = (uint32)_textureCache.size();
 
-    return &(cacheItr.first->second);
+    CachedTextureInfo info = LoadImageTexture(image);
+    info.image = image;
+
+    _textureCache.push_back(info);
```

Comment:
You are still abusing object's lifetimes.

Let's assume the vector's capacity doubles each time it runs out of space.

https://github.com/OpenRCT2/OpenRCT2/blob/15d0a726ac9f49470021615dae785db6f2ebad1c/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp#L712-L713

When you start with cold cache, you will have extended the vector capacity to 1 and you return a const-reference to a newly allocated, inserted object. The second line calls for next object, but as the vector is out of space, it reallocates to increase capacity to 2. You return the new object, but the previously returned `textureMask` points at a dead storage, as `push_back` will have invalidated it.

Reply:
Good point, I'll think of something.